content/blog/performant-code/index.Rmd

```diff
@@ -0,0 +1,526 @@
+```{r bm-relocate}
+bench::mark(
+  mutate = mutate(mtcars_tbl, year = 1974L, .after = make_model),
+  relocate = relocate(mtcars_tbl, year, .after = make_model),
+  `[.tbl_df` = 
+      mtcars_tbl[
+        c(left_cols, 
+          colnames(mtcars_tbl[!colnames(mtcars_tbl) %in% left_cols])
+        )
+      ],
+  check = FALSE
+) %>% 
+  select(expression, median)
+```
+
+```{r null-year, include = FALSE}
+mtcars_tbl$year <- NULL
+```
+
+Some examples of changes like this made to tidymodels packages: [tidymodels/parsnip#935](https://github.com/tidymodels/parsnip/pull/935), [tidymodels/parsnip#933](https://github.com/tidymodels/parsnip/pull/933), [tidymodels/parsnip#921](https://github.com/tidymodels/parsnip/pull/921), and [tidymodels/tune#635](https://github.com/tidymodels/tune/pull/635).
+
+#### `pull()`
+
+The dplyr code:
+
+```{r pull, eval = FALSE}
+pull(mtcars_tbl, hp)
+```
+
+...can be replaced by:
+
+```{r pull-alt, eval = FALSE}
+mtcars_tbl$hp
+```
+
+...or:
+
+```{r pull-alt-2, eval = FALSE}
+mtcars_tbl[["hp"]]
+```
+
+Note that `$` will raise a warning if the subsetted column doesn't exist, while `[[` will silently return `NULL`.
+
+With benchmarks:
+
+```{r bm-pull}
+bench::mark(
+  dplyr      = pull(mtcars_tbl, hp),
+  `$.tbl_df` = mtcars_tbl$hp,
+  `[[.tbl_df` = mtcars_tbl[["hp"]]
+) %>%
+  select(expression, median)
+```
+
+Some examples of changes like this made to tidymodels packages: [tidymodels/parsnip#935](https://github.com/tidymodels/parsnip/pull/935) and [tidymodels/tune#635](https://github.com/tidymodels/tune/pull/635).
+
+#### `bind_*()`
+
+`bind_rows()` and `bind_cols()` can be substituted for `vec_rbind()` and `vec_cbind()`, respectively. First, row-binding:
+
+```{r bm-bind-rows}
+bench::mark(
+  dplyr = bind_rows(mtcars_tbl, mtcars_tbl),
+  vctrs = vec_rbind(mtcars_tbl, mtcars_tbl)
+) %>%
+  select(expression, median)
+```
+
+As for column-binding:
+
+```{r bm-bind-cols}
+tbl <- tibble(year = rep(1974L, nrow(mtcars_tbl)))
+
+bench::mark(
+  dplyr = bind_cols(mtcars_tbl, tbl),
+  vctrs = vec_cbind(mtcars_tbl, tbl)
+) %>%
+  select(expression, median)
+```
+
+Some examples of changes like this made to tidymodels packages: [tidymodels/tune#636](https://github.com/tidymodels/tune/pull/636).
+
+#### `group_by()` + `summarize()`
+
+In general, the introduction of groups makes these substitutions much trickier. In those cases, it's likely best to weigh (via profiling) how significant the slowdown is and, if it's not too bad, opt not to make any changes. For code that relies on `group_by()` and sees heavy traffic, see `vctrs::list_unchop()`, `vctrs::vec_chop()`, and `vctrs::vec_rep_each()`.
+
+### Tibbles
+
+Tibbles are great, and I don't want to interface with any other data frame-y thing. Some notes:
+
+- `as_tibble()` on a tibble is not "free":
+
+```{r bm-as-tibble}
+bench::mark(
+  on_tbl_df = as_tibble(mtcars_tbl),
+  on_data.frame = as_tibble(mtcars, rownames = "make_model")
+) %>% 
+  select(expression, median)
+```
```

Comment:
I believe you can get the block aligned with the above bullet if you add 3 spaces to the beginning of each line in the block, including the triple backtick lines